src/rewritable_units/mutations.rs

```diff
@@ -27,68 +32,174 @@ pub(super) fn content_to_bytes(
-            self.content_after.splice(pos..pos, c.iter().cloned());
+    #[inline]
+    pub const fn removed(&self) -> bool {
+        self.removed
+    }
+}
 
-            pos += c.len();
-        });
+impl From<(&str, ContentType)> for StringChunk {
+    #[inline]
+    fn from((content, content_type): (&str, ContentType)) -> Self {
+        Self::Buffer(Box::from(content), content_type)
     }
+}
 
+pub(crate) enum StringChunk {
+    Buffer(Box<str>, ContentType),
+    Stream(Box<dyn StreamingHandler>),
+}
+
+#[derive(Default)]
+pub(crate) struct DynamicString {
+    chunks: Vec<StringChunk>,
+}
+
+impl DynamicString {
     #[inline]
-    pub fn replace(&mut self, content: &str, content_type: ContentType) {
-        let mut replacement = Vec::default();
+    pub const fn new() -> Self {
+        Self { chunks: vec![] }
+    }
 
-        content_to_bytes(content, content_type, self.encoding, &mut |c| {
-            replacement.extend_from_slice(c);
-        });
+    #[inline]
+    pub fn clear(&mut self) {
+        self.chunks.clear();
+    }
 
-        self.replacement = replacement;
-        self.remove();
+    #[inline]
+    pub fn push_front(&mut self, chunk: StringChunk) {
+        self.chunks.insert(0, chunk);
     }
 
     #[inline]
-    pub fn remove(&mut self) {
-        self.removed = true;
+    pub fn push_back(&mut self, chunk: StringChunk) {
+        self.chunks.push(chunk);
     }
 
+    pub fn into_bytes(
+        self,
+        encoding: &'static Encoding,
+        output_handler: &mut dyn FnMut(&[u8]),
+    ) -> BoxResult {
+        for chunk in self.chunks {
+            match chunk {
+                StringChunk::Buffer(content, content_type) => {
+                    content_to_bytes(&content, content_type, encoding, output_handler);
+                }
+                StringChunk::Stream(handler) => {
+                    let output_handler = &mut *output_handler;
+                    handler.write_all(&mut StreamingHandlerSink {
+                        encoding,
+                        output_handler,
+                        _marker: PhantomData,
+                    })?;
+                }
+            };
+        }
+        Ok(())
+    }
+}
+
+/// Used to write chunks of text or markup in streaming mutation handlers.
+///
+/// Argument to [`StreamingHandler::write_all`]
+pub struct StreamingHandlerSink<'tmp> {
+    encoding: &'static Encoding,
+    output_handler: &'tmp mut dyn FnMut(&[u8]),
+    /// Proactively mark as !Send !Sync in case it needs it later
+    _marker: PhantomData<*const ()>,
+}
+
+impl<'tmp> StreamingHandlerSink<'tmp> {
+    /// Writes the given UTF-8 string to the output, converting the encoding and [escaping](ContentType) if necessary.
+    ///
+    /// It may be called multiple times. The strings will be concatenated together.
     #[inline]
-    pub fn removed(&self) -> bool {
-        self.removed
+    pub fn write_str_chunk(&mut self, content: &str, content_type: ContentType) {
```

Comment:
For workers use cases they want to write byte slices which makes the whole thing more complex. So, we would need streaming encoding of the written chunks

Reply:
The existing API only accepts UTF-8, and Workers use that API, so I assume Workers have UTF-8 to send, and it's not a case of any-encoding to any-encoding conversion.

UTF-8 data split into arbitrary chunks just needs finding a char boundary at the end and buffering up to 3 bytes, so this can be added without major changes to the API.

Reply:
correct. we only need utf-8 support as that's what is used by workers and we are merging streams produced by workers/pages.